test: optimise test with chai custom method overwrite

MaikoTan opened the pull request.

MaikoTan committed:
test: optimise test with chai custom method overwrite

MaikoTan committed:
refactor: do a little formatting

SaarChaffee commented on __tests__/index.spec.ts:
```diff
@@ -1,30 +1,48 @@
+}
+
+Assertion.overwriteChainableMethod('a', a, noop)
+Assertion.overwriteChainableMethod('an', a, noop)
```
```suggestion
import { Assertion, expect, util } from 'chai'

import { defaultRules, getNews, LodestoneNews, regions, News } from '../src'

const UTC_0_2022_07_19_07_59_59 = 1658217599 * 1000 // 2022-07-19 07:59:59 UTC+0

Assertion.addProperty('newsList', function () {
  new Assertion(this._obj).to.be.instanceof(Array<News>)
  this._obj.forEach(n => {
    new Assertion(n).to.have.deep.property('title').which.is.a('string')
    new Assertion(n).to.have.deep.property('epoch').which.is.a('number')
    new Assertion(n).to.have.deep.property('url').which.is.a('string')
    new Assertion(n).to.have.deep.property('date').which.is.a('date')
  })
  this.assert(
    this._obj instanceof Array<News>,
    'expected #{this} to be an array of News',
    'expected #{this} to not be an array of News',
    null,
  )
})

Assertion.addMethod('haveLen', function (n?: number) {
  if (util.flag(this, 'newsList')) {
    new Assertion(this._obj).to.have.deep.property('length').which.is.a('number')
    if (n) {
      new Assertion(this._obj).to.have.lengthOf(n)
      this.assert(
        this._obj.length === n,
        'expected #{this} to have #{exp} news, but got #{act}',
        'expected #{this} to not have #{exp} news, but got #{act}',
        n,
        this._obj.length,
      )
    } else {
      new Assertion(this._obj).to.have.lengthOf.greaterThan(0)
      this.assert(
        this._obj.length > 0,
        'expected #{this} to have more than one news, but got #{exp}',
        'expected #{this} to not have more than one news, but got #{exp}',
        this._obj.length,
      )
    }
  }
})
```

SaarChaffee commented on __tests__/index.spec.ts:
```diff
@@ -1,30 +1,48 @@
+        })
+        expect(news).is.an('array')
+        expect(news).has.length.greaterThan(0)
+        expect(news[0]).is.a('news')
```
现在可以这么用，`haveLen()` 入参为空则判断大于0
```suggestion
        expect(news).is.a.newsList.that.haveLen()
```

SaarChaffee commented on __tests__/index.spec.ts:
```diff
@@ -83,9 +95,6 @@ it('should filter with after', async () => {
-  expect(news[0]).has.property('url').which.is.a('string')
-  expect(news[0]).has.property('date').which.is.a('date')
-  expect(news[news.length - 1].date).is.greaterThan(new Date(UTC_0_2022_07_19_07_59_59))
+  expect(news[0]).is.a('news')
```
```suggestion
  expect(news).is.a.newsList.that.haveLen()
```

SaarChaffee commented on __tests__/index.spec.ts:
```diff
@@ -51,10 +69,7 @@ it('should accept custom config', async () => {
-  expect(news[0]).has.property('epoch').which.is.a('number')
-  expect(news[0]).has.property('url').which.is.a('string')
-  expect(news[0]).has.property('date').which.is.a('date')
+  expect(news[0]).is.a('news')
```
```suggestion
  expect(news).is.a.newsList.that.haveLen(10)
```

SaarChaffee commented on __tests__/index.spec.ts:
```diff
@@ -66,11 +81,8 @@ it('should filter with after / before', async () => {
-  expect(news[0]).has.property('date').which.is.a('date')
-  expect(news.length).is.equal(2)
+  expect(news[0]).to.be.a('news')
+  expect(news).has.length(2)
```
```suggestion
  expect(news).is.a.newsList.that.haveLen(2)
```

MaikoTan committed:
fix: change value type as unknown

MaikoTan committed:
test: fix length and add newsList overwrite

MaikoTan committed:
test: add chai-things for all assertion

SaarChaffee commented on __tests__/index.spec.ts:
```diff
@@ -29,12 +40,10 @@ Assertion.overwriteChainableMethod('an', a, noop)
+  new Assertion(this._obj).all.have.property('epoch').which.is.a('number')
+  new Assertion(this._obj).all.have.property('url').which.is.a('string')
+  new Assertion(this._obj).all.have.property('date').which.is.a('date')
 })
```
```suggestion
  util.flag(this, 'newsList', true)
})
```

SaarChaffee commented on __tests__/index.spec.ts:
```diff
@@ -27,6 +27,40 @@ function noop() {
+        new Assertion(this._obj).to.have.length(n, message)
+      } else {
+        new Assertion(this._obj).to.have.length.greaterThan(0, message)
+      }
```
```suggestion
      }
      util.flag(this, 'newsList', false)
```

MaikoTan committed:
Apply suggestions from code review

Co-authored-by: 霞飛 <53519746+SaarChaffee@users.noreply.github.com>

MaikoTan committed:
Merge branch 'master' into optimise-test

MaikoTan committed:
test: use chai-each instead of chai-things

as it plays well in type, so that typescript would not be
urusai on these annoying things

MaikoTan committed:
test: fix typing & simplify some test assertion

MaikoTan committed:
test: fix should language chain

MaikoTan committed:
test: fix each chain for news list

MaikoTan committed:
test: revert comment

MaikoTan commented:
[The lodestone is maintaining currently](https://jp.finalfantasyxiv.com/lodestone/news/detail/5bcbc0d889444667265d97d324205537b508973b), which broke the test always.

SaarChaffee commented:
```log
 29 passing (3m)
 1 failing
```

MaikoTan commented:
you can trigger the ci test by clicking the re-run button.

SaarChaffee commented on __tests__/index.spec.ts:
```diff
@@ -49,12 +99,7 @@ it('should accept custom config', async () => {
-  expect(news[0]).has.property('epoch').which.is.a('number')
-  expect(news[0]).has.property('url').which.is.a('string')
-  expect(news[0]).has.property('date').which.is.a('date')
+  expect(news).to.be.newsList.which.has.length(20)
```
```suggestion
  expect(news).to.be.newsList.which.has.length(10)
```

MaikoTan committed:
Update __tests__/index.spec.ts

Co-authored-by: 霞飛 <53519746+SaarChaffee@users.noreply.github.com>
